data_model/src/predicate.rs

```diff
@@ -2,54 +2,177 @@
+            tail: vec![$($t),*],
+        }
+    }};
+}
+
 /// Predicate combinator enum.
 #[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
-pub enum PredicateBox {
-    /// Logically `&&` the results of applying the two predicates.
-    And(Vec<PredicateBox>),
-    /// Logically `||` the results of applying the two predicates.
-    Or(Vec<PredicateBox>),
+// Ideally we would enforce `P: PredicateTrait<Input>` here, but I
+// couldn't find a way to do it without polluting everything
+// downstream with explicit lifetimes, since we would need to
+// store PhantomData<Input> here, and `Input`s are most often
+// references (e.g. &Value).
+pub enum GenericPredicateBox<P> {
+    /// Logically `&&` the results of applying the predicates.
+    And(NonEmpty<Self>),
+    /// Logically `||` the results of applying the predicates.
+    Or(NonEmpty<Self>),
     /// Negate the result of applying the predicate.
-    Not(Box<PredicateBox>),
+    Not(Box<Self>),
     /// The raw predicate that must be applied.
-    Raw(value::Predicate),
+    Raw(P),
 }
 
-impl PredicateBox {
+impl<P> Display for GenericPredicateBox<P>
+where
+    P: Display,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            GenericPredicateBox::And(predicates) => {
+                write!(f, "AND(")?;
+                predicates.head.fmt(f)?;
+                for predicate in &predicates.tail {
+                    predicate.fmt(f)?;
+                }
+                write!(f, ")")
+            }
+            GenericPredicateBox::Or(predicates) => {
+                write!(f, "OR(")?;
+                predicates.head.fmt(f)?;
+                for predicate in &predicates.tail {
+                    predicate.fmt(f)?;
+                }
+                write!(f, ")")
+            }
+            GenericPredicateBox::Not(predicate) => write!(f, "NOT({})", predicate),
+            GenericPredicateBox::Raw(predicate) => predicate.fmt(f),
+        }
+    }
+}
+
+impl<P> GenericPredicateBox<P> {
     /// Construct [`PredicateBox::Raw`] variant.
     #[inline]
-    pub fn new(pred: impl Into<value::Predicate>) -> Self {
+    pub fn new<Input>(pred: impl Into<P>) -> Self
+    where
+        P: PredicateTrait<Input>,
+        Input: Copy,
+    {
         Self::Raw(pred.into())
     }
 
     /// Construct [`PredicateBox::And`] variant.
     #[inline]
-    pub fn and(left: impl Into<PredicateBox>, right: impl Into<PredicateBox>) -> Self {
-        Self::And(vec![left.into(), right.into()])
+    pub fn and(left: impl Into<Self>, right: impl Into<Self>) -> Self {
+        match (left.into(), right.into()) {
+            (Self::And(mut left), Self::And(right)) => {
+                left.tail.push(*right.head);
+                left.tail.extend(right.tail);
+                Self::And(left)
+            }
+            (Self::And(mut and), other) | (other, Self::And(mut and)) => {
+                and.tail.push(other);
```

Comment:
We should be careful with reordering of predicates.
Operation is commutative so result would be the same but we can degrade early exit accidentally.

Reply:
True, another level of pointer indirection should be a better trade-off than degraded early exit. I removed out-of-order matches here.